From 29a13422b5f855b2641373a7cb238a20c8a97cc0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?De=C3=A9=20G=C3=A1bor?= <dee.gabor@gmail.com>
Date: Sun, 4 Jan 2026 16:45:39 +0100
Subject: [PATCH 089/210] WIP: arm64: dts: amlogic: s905d-vero4k-plus: fix
 ethernet configuration
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Align the ethmac and external_mdio nodes with other S905D (Gbit)
devices and following the mc_val_external_phy value used in the
original vendor kernel dts [0] and changes to the vendor driver
[1] set tx-delay-ns to 4ns to address performance and stability.

[0] https://github.com/osmc/vero3-linux/blob/master/arch/arm64/boot/dts/amlogic/vero3plus_2g_16g.dts
[1] https://github.com/osmc/vero3-linux/commit/376838b896e48fcdb5129735ff0bfb5f14b0210b

Fixes: d926a4fe2533 ("arm64: dts: meson: add support for OSMC Vero 4K+")
Signed-off-by: Deé Gábor <dee.gabor@gmail.com>
Signed-off-by: Christian Hewitt <christianshewitt@gmail.com>
---
Checkpatch generates the 'phy-mode "rgmii" without comment'
warning but this dts is now using the same rgmii format as
most other Amlogic boards, so I have omitted the comment.

The default delay used with other S905D and S912 board dts
is 2ns but iperf3 experiments show the vero4k-plus board
needs a longer 4ns delay to function correctly:

Testing with 2ns:

LibreELEC:~ # iperf3 -c 10.60.0.30
Connecting to host 10.60.0.30, port 5201
[  5] local 10.60.0.51 port 41706 connected to 10.60.0.30 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  57.1 MBytes   479 Mbits/sec   68   28.5 KBytes
[  5]   1.00-2.00   sec  65.0 MBytes   545 Mbits/sec   66   67.0 KBytes
[  5]   2.00-3.00   sec  64.0 MBytes   537 Mbits/sec   73   27.1 KBytes
[  5]   3.00-4.00   sec  61.5 MBytes   516 Mbits/sec   69   25.7 KBytes
[  5]   4.00-5.00   sec  62.5 MBytes   524 Mbits/sec   75   28.5 KBytes
[  5]   5.00-6.00   sec  57.6 MBytes   483 Mbits/sec   63   14.3 KBytes
[  5]   6.00-7.00   sec  59.8 MBytes   501 Mbits/sec   64   29.9 KBytes
[  5]   7.00-8.00   sec  64.4 MBytes   540 Mbits/sec   67   48.5 KBytes
[  5]   8.00-9.00   sec  63.4 MBytes   532 Mbits/sec   69    120 KBytes
[  5]   9.00-10.00  sec  55.8 MBytes   467 Mbits/sec   63   52.8 KBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec   611 MBytes   512 Mbits/sec  677   sender
[  5]   0.00-10.01  sec   610 MBytes   512 Mbits/sec        receiver

Testing with 4ns:

LibreELEC:~ # iperf3 -c 10.60.0.30
Connecting to host 10.60.0.30, port 5201
[  5] local 10.60.0.51 port 50932 connected to 10.60.0.30 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec   114 MBytes   955 Mbits/sec    0    339 KBytes
[  5]   1.00-2.00   sec   113 MBytes   948 Mbits/sec    0    339 KBytes
[  5]   2.00-3.00   sec   113 MBytes   946 Mbits/sec    0    409 KBytes
[  5]   3.00-4.00   sec   113 MBytes   950 Mbits/sec    0    409 KBytes
[  5]   4.00-5.00   sec   112 MBytes   944 Mbits/sec    0    409 KBytes
[  5]   5.00-6.00   sec   113 MBytes   948 Mbits/sec    0    409 KBytes
[  5]   6.00-7.00   sec   112 MBytes   942 Mbits/sec    0    409 KBytes
[  5]   7.00-8.00   sec   113 MBytes   949 Mbits/sec    0    409 KBytes
[  5]   8.00-9.00   sec   112 MBytes   943 Mbits/sec    0    409 KBytes
[  5]   9.00-10.00  sec   113 MBytes   944 Mbits/sec    0    409 KBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  1.10 GBytes   947 Mbits/sec    0    sender
[  5]   0.00-10.00  sec  1.10 GBytes   946 Mbits/sec         receiver
---
 .../amlogic/meson-gxl-s905d-vero4k-plus.dts   | 20 ++++---------------
 1 file changed, 4 insertions(+), 16 deletions(-)

diff --git a/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-vero4k-plus.dts b/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-vero4k-plus.dts
index 595b49085074..57807acb1871 100644
--- a/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-vero4k-plus.dts
+++ b/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-vero4k-plus.dts
@@ -42,18 +42,15 @@ &ethmac {
 	pinctrl-0 = <&eth_pins>;
 	pinctrl-names = "default";
 
-	phy-mode = "rgmii-txid";
+	phy-mode = "rgmii";
 	phy-handle = <&external_phy>;
 
-	amlogic,tx-delay-ns = <0>;
+	amlogic,tx-delay-ns = <4>;
 };
 
 &external_mdio {
 	external_phy: ethernet-phy@0 {
 		/* Realtek RTL8211F (0x001cc916) */
-		pinctrl-0 = <&eth_phy_irq_pin>;
-		pinctrl-names = "default";
-
 		reg = <0>;
 		max-speed = <1000>;
 
@@ -62,19 +59,10 @@ external_phy: ethernet-phy@0 {
 		reset-gpios = <&gpio GPIOZ_14 GPIO_ACTIVE_LOW>;
 
 		interrupt-parent = <&gpio_intc>;
+		/* MAC_INTR on GPIOZ_15 */
 		interrupts = <25 IRQ_TYPE_LEVEL_LOW>;
-	};
-};
 
-&pinctrl_periphs {
-	/* Ensure the phy irq pin is properly configured as input */
-	eth_phy_irq_pin: eth-phy-irq {
-		mux {
-			groups = "GPIOZ_15";
-			function = "gpio_periphs";
-			bias-disable;
-			output-disable;
-		};
+		eee-broken-1000t;
 	};
 };
 
-- 
2.43.0

